src/rulesets/Base/array.jl

```diff
@@ -278,3 +278,128 @@ function rrule(::typeof(fill), value::Any, dims::Int...)
     end
     return fill(value, dims), fill_pullback
 end
+
+#####
+##### `findmax`, `maximum`, etc.
+#####
+
+for findm in (:findmin, :findmax)
+    findm_pullback = Symbol(findm, :_pullback)
+
+    @eval function frule((_, xdot), ::typeof($findm), x; dims=:)
+        y, ind = $findm(x; dims=dims)
+        return (y, ind), Tangent{typeof((y, ind))}(xdot[ind], NoTangent())
+    end
+
+    @eval function rrule(::typeof($findm), x::AbstractArray{<:Number}; dims=:)
+        y, ind = $findm(x; dims=dims)
+        project = ProjectTo(x)
+        function $findm_pullback((dy, _))
+            x_thunk = @thunk begin
+                dx = fill!(similar(x, eltype(dy)), false)
+                view(dx, ind) .= dy  # possibly 0-dim view, allows dy::Number and dy::Array, and dx::CuArray
+                project(dx)
+            end
+            x_ithunk = InplaceableThunk(x_thunk) do dx
+                view(dx, ind) .= view(dx, ind) .+ dy  # this could be .+=, but not on Julia 1.0
+                dx
+            end
+            return (NoTangent(), x_ithunk)
+        end
+        function $findm_pullback(::Tuple{AbstractZero, Any})
+            return (NoTangent(), NoTangent())
+        end
+        return (y, ind), $findm_pullback
+    end
+
+end
+
+# These rules for maximum pick the same subgradient as findmax:
+
+function frule((_, xdot), ::typeof(maximum), x; dims=:)
+    y, ind = findmax(x; dims=dims)
+    return y, xdot[ind]
+end
+
+function rrule(::typeof(maximum), x::AbstractArray{<:Number}; dims=:)
+    (y, _), back = rrule(findmax, x; dims=dims)
+    maximum_pullback(dy) = back((dy, nothing))
+    return y, maximum_pullback
+end
+
+function frule((_, xdot), ::typeof(minimum), x; dims=:)
+    y, ind = findmin(x; dims=dims)
+    return y, xdot[ind]
+end
+
+function rrule(::typeof(minimum), x::AbstractArray{<:Number}; dims=:)
+    (y, _), back = rrule(findmin, x; dims=dims)
+    minimum_pullback(dy) = back((dy, nothing))
+    return y, minimum_pullback
+end
+
+#####
+##### `extrema`
+#####
+
+function rrule(::typeof(extrema), x::AbstractArray{<:Number}; dims=:)
```

Comment:
similar comments apply here, i.e. accepting `Tangent` and reusing `getindex` rrule.  Maybe we can even get rid of special casing the colon and cartesian indices this way?

Reply:
The special-casing is partly an unzip thing -- `extrema(rand(3,3), dims=1)` turns out to return not a tuple of min,max arrays, but the other way around.